File: leetcode/java/347.top-k-frequent-elements.java
class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int num : nums){
            map.put(num, map.getOrDefault(num, 0)+1);
        }
        List<Map.Entry<Integer, Integer>> list = new ArrayList<>(map.entrySet());
        list.sort(Map.Entry.comparingByValue());
        List<Integer> res = new ArrayList<>();
        while(k > 0){
            res.add(list.get(list.size() - k).getKey());
            k--;
        }
        return res;
    }
}

File: leetcode/golang/347.top-k-frequent-elements.go
func topKFrequent(nums []int, k int) []int {

	counters := make(map[int]int)
	for _, v := range nums {
		counters[v]++
	}

	feqs := make([][]int, len(nums)+1)
	for i := range feqs {
		feqs[i] = make([]int, 0)
	}

	for k, v := range counters {
		feqs[v] = append(feqs[v], k)
	}

	res := make([]int, 0)
	j := k
	for i := len(nums); i >= 0 && j > 0; i-- {
		if len(feqs[i]) != 0 {
			for _, v := range feqs[i] {
				res = append(res, v)
				j--
				if len(res) == k {
					return res
				}
			}
		}
	}

	return res
}